### DISCUSSION TITLE
RandomLinkSplit Abilities / Suggestions

### DISCUSSION THREAD

**Discussion by Guykor** [pyg-team/pytorch_geometric#7828] [Q&A]
Hello :)
**Usage:**
I'm using the splitter on a KG which I treat as a homogenous graph, where edge types are added as edge_attr [num_edges, 1] to the Data object (where each edge type is an integer).
My GNN ignores node types (all have the same embedding dim), but I use Relation-wise weight in link prediction (DistMult).
I've encountered some difficulties using RandomLinkSplit that I think force me to implement a simpler version of my own, but maybe I'm wrong and these abilities are already implemented and I'm missing something.
**Required abilities**
1. edge_attr for the supervision (label) edges - how can I query from test_data (Data object) the attribute of the pos and neg edges for DistMult?
2. Custom negative sampler - maybe there is a way to pass a custom sampler function?
3. Seed randomness - to reproduce results. 

The same considerations (except seeding) apply in LinkNeighborLoader - I don't have the label edges attributes, so I can't pass triplet neg sampling when I choose a sampling strategy.
 
I've read a lot of documentation and tutorials, but I'm still unsure if I can use existing tools out of the box for my task

Thank you for your amazing work!

**Comment by rusty1s** [comment]
1. Can you clarify what you mean? Negative labels will be given as zeros in the `edge_label` tensor. What do you want to query in particular? You should be also able to just do `pos_edge_label, neg_edge_label = edge_label.split(2)`.
2. There does not exist such a function yet, sorry :(
3. You should be able to do this via `torch.manual_seed(xxx)`.

Overall, I agree with your concerns though. Often times, it is easier to implement splitting on your own since it is seems to be impossible to cover all use-cases with a general splitter. The interface to `RandomLinkSplit` is already quite complex, and I would like to avoid to make it even more complex.

**Reply by Guykor** [reply]
1. I mean that when I use negative sampling from within RandomLinkSplit (for validation and test edges) - the sampling ignores edge types (which I referred to as 'attr'). Hence, supervision edge type is not available, and using something like DistMult becomes impossible with this implementation.

I saw some issues regarding plans supporting more heterogenous cases (for nodes and types) but it still seems in progress.

**Reply by rusty1s** [reply]
I think you can still do something like DistMult, but it requires a homogeneous graph where edge type information is passed as an edge-level vector (e.g., via `data.to_homogeneous()`). Would that work for you?
